SI-7349 SI-7358 SI-7409 Par-Test checks and balances

som-snytt opened the pull request:
A failing scalacheck test generates a test failure instead of allowing the build to succeed.
The failure log is added to the transcript.

Updates Scalacheck to 1.10.1 and invokes it using the
test-interface API. This obviates kludgy output checking for result
status.

Also, Partest testnum field width is sensitive to total tests.

Allow the hyphens to line up on a column within the output
for a test category or "kind".

Users are notified when check files are updated, and also when the compiler crashes instead of updating their check files.

Supersedes https://github.com/scala/scala/pull/2418

Review by @soc @adriaanm @retronym

som-snytt committed:
SI-7358 Partest fails on scalacheck failure

A failing test generates a test failure.
The failure log is added to the transcript.
The log stream is closed when done.

som-snytt committed:
Partest testnum field width is sensitive to total tests

Allow the hyphens to line up on a column within the output
for a test category or "kind".

```
$ partest files/neg/t4134.scala --run
Selected 1348 tests drawn from 1 named test categories, specified tests

& starting 1 test in neg
ok 1 - neg/t4134.scala

& starting 1347 tests in run
ok    1 - run/absoverride.scala
ok    2 - run/amp.scala
ok    3 - run/adding-growing-set.scala
```

som-snytt committed:
Partest has an optionable wait period.

Partest --timeout "30 seconds" to time out the test run.

It will not hang on timeout ("I hang with Par-Test" t-shirts
not-withstanding).

It's beyond the scope of this commit to investigate argument
parsing:

`partest --timeout "\"30 seconds"\" --pos`

som-snytt committed:
SI-7409 Par-Test: A crash is not a DNC for neg tests

A compiler crash does not count as Does Not Compile for
purposes of a negative test. Changing the test kind from
"neg" to "dnc" is out of scope for this PR.

Even if the user asks to update the check file with the
crash result, we must prevent him from doing so. Any
further attempts to update the check file with a crash
will dispatch the Scala SWAT squad which will race to
his location and physically restrain the user. Only
Martin holds the code which will allow the squad to
stand down. So make sure his cell is on next time you
want to --update-check.

A neg test will stop trying to compile after the first
failed round, which is all that matters.

By popular request, a new test outcome is emitted when
the check file is updated. It is called "Updated". It
even has its own short status, the double-plus that is
reminiscent of diff output.

som-snytt committed:
Update ScalaCheck to 1.10.1.

Puts test-interface on the class path.

som-snytt committed:
SI-7349 Partest supports test-interface

Partest uses test-interface API to invoke ScalaCheck.  This obviates
ad hoc output checking for result status.

The context class loader is set to a loader that the scaladoc scalacheck
tests can use.

soc commented on src/partest/scala/tools/partest/PartestDefaults.scala:
```diff
@@ -21,6 +22,7 @@ object PartestDefaults {
+  def waitTime   = propOrNone("partest.timeout") map (Duration.apply) getOrElse Duration("4 hours")
 
-  def timeout     = "1200000"
+  //def timeout     = "1200000"   // per-test timeout
```
Is this an oversight? Be careful, committing deactivated pieces of code can enrage the merge gods. :-)

som-snytt replied:
It will be reactivated someday.  Don't make me regret burning --timeout for the global timeout.

Maybe --fail-after.

```
$ fastest --help
Usage: partest [<options>] [<testfiles>]

  To control test output:
    --help         show this message
    --version      show version
    --debug        enable debug output
    --verbose      show progress
    --summary      summary results only
    --statistics   would you like some numbers with that?
    --silent       show nothing
    --show-log     show log output on error
    --show-diff    show diff between log and check file
    --colors       do it with style
    --gui          do it with swing

  To select libraries and sources for testing:
    --build        where to find lib/*.jar
    --classes      where to find classes
    --pack         selects --build=build/pack
    --quick        selects --classes=build/quick/classes
    --installed    selects --build=./lib
    --dist         selects --build=dists/latest/lib
    --latest       detect the latest from quick, pack, installed, dist
    --sourcepath   set path to test sources

  To select tests to run:
    --all          run all tests
    --failed       run previously failed tests
    --aggregate    compile tests in aggregate
    --pos          Testing compiler (on files whose compilation should succeed)
    --neg          Testing compiler (on files whose compilation should fail)
    --run          Testing interpreter and backend
    --jvm          Testing JVM backend
    --res          Testing resident compiler
    --fsc          Testing compile server (on files whose compilation should succeed)
    --buildmanager Testing Build Manager
    --shootout     Testing shootout tests
    --script       Testing script tests
    --scalacheck   Running ScalaCheck tests
    --scalap       Running scalap decompiler tests
    --specialized  Testing specialization
    --presentation Testing presentation compiler
    --instrumented Running instrumented tests
    --ant          Testing Ant task
    --grep         run all tests with any file containing <expr>
    --skip         skip tests selected by <expr>
    --ticket       run tests for ticket #<NNNN>

  Options for test in development:
    --list         list tests that would run (dry run)
    --limit        abort after limit failures
    --tolerance    succeed on up to tolerance failures
    --ignore       ignore missing or bad tests
    --timeout      timeout to fail a test, as a duration (default: 10 minutes)
    --cleaning     on success, delete artifacts from previous failure
    --update       on failure, update check files

  To select how to run the compilers and tests:
    --fork         fork compilation processes
    --forking      fork [scalac, javac, tests]
    --reuse        reuse compiler instances
    --disk         compiler inputs and outputs on disk
    --cachecp      reduce io on class path access
    --par          parallelism

```

som-snytt committed:
Partest redirects stderr to log file

Some scalac output is on stderr, and it's useful to see that
in the log file, especially for debugging.

soc commented:
PLS REBUILD ALL

soc commented:
Any chance to get this merged? The `--timeout` option is crucial for running the test-suite on Avian (without having to manually modify the partest sources).

paulp commented:
There are failing tests.

som-snytt commented:
I added stderr to logs; when I have a chance later, I'll set flags or update checks.  Today I'm chaperoning kindergarten to the science museum. Hey, I thought this PR was months from a merge, if ever.  Or, take it without the last commit.

scala-jenkins commented:
(kitty-note-to-self: ignore 17169810)
:cat: Roger! Rebuilding pr-rangepos-per-commit, pr-checkin-per-commit for 87f292dd, 8557ceed, a09545ee, 1f57d309, 7f4d694e, b65b537e, 7221de3a. :rotating_light:

soc commented:
Ah, right. I didn't see the last, failing commit and thought that the one before was just stuck.

paulp commented:
How can you add stderr to logs without breaking all the tests? Does it not compare the log to the checkfile?

som-snytt commented:
Answering rhetorically, I expected somewhere between some and all tests to break until updated.  But if there are tests that really don't like stderr, it would have to go behind a flag or -D.  I needed the feature to see stderr interleaved with stdout for debugging purposes. I could have stashed it in a branch.

som-snytt commented:
Superseded in turn.  Logged stderr will come to a PR near you soon.  Or maybe direct to video.
